AI #814: Revised spec development guidelines

mike-finopsorg opened the pull request:
Added a collection of processes we are using in the FOCUS project for spec development. 

Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
added initial version of the way we work document to the guidelines folder
Signed-off-by: Mike Fuller <mike@finops.org>

shawnalpay commented on guidelines/development-processes.md:
```diff
@@ -0,0 +1,292 @@
+
+Our branch promotion process takes the following path:
+
+- _<\<dev\_branch>>:_ The feature branch in which a FOCUS Working Group member creates changes.  Such feature branches are ephemeral and are automatically deleted once their content is merged into the working draft.
+
+- _working\_draft:_ The branch in which all approved changes are collected during development.
+
+- _candidate\_recommendation_: The branch in which all approved changes are collected at the end of development for a release. Content in this branch is held for 30 days to enable the [Intellectual Property Rights (IPR) review process](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/ipr.md).
+
+- _main_: The branch in which all approved changes are published after Steering Committee ratification of the latest release.
```
@mike-finopsorg Were these spaces between bullets intentional or not?  Wasn't sure if this was from the copy/paste from Google Docs or if you wanted them this way.

Same question on the escape characters.  Intentional or not?

shawnalpay commented on guidelines/development-processes.md:
I believe this document should replace and absorb the content in [spec-development-guidelines.md](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/guidelines/spec-development-guidelines.md).  Having these two documents next to each other is confusing IMO.

mike-finopsorg replied:
I renamed spec-development-guidelines.md to development-tips.md for now as I do not have time to refactor that content into this file. This keeps the content available but moves it to a less priority filename. If someone has time to merge the content into this PR I would be happy to remove the tips file and accept the suggestion.

shawnalpay replied:
I do think that old document is out of date (e.g., it references a Google Doc that has been created for the issue, which is no longer how we work).  We need to either fold it in or jettison it.

shawnalpay replied:
@mike-finopsorg I added a section called `Content Creation` that folds in the only section from `development-tips.md` that I thought was relevant.  I think we can delete that markdown now.  Let's sync and align on that.

jpradocueva commented on guidelines/development-processes.md:
```diff
@@ -0,0 +1,292 @@
+
+### Pull Request Approval and Conflict Resolution
+
+All Pull Requests must go through a formal review and approval process before being merged into the  `working_draft` branch.
```
Shall we write down the rule for presenting the PR for review and approval the following week?

mike-finopsorg replied:
Same here, please suggest edits

jpradocueva commented on guidelines/development-processes.md:
```diff
@@ -0,0 +1,292 @@
+All FOCUS members are encouraged to review and make comments and suggestions. With the aim for all changes to the specification to be by consensus. 
+
+
+### Resolving Comments and Suggestions
```
I suggest linking this section with the "Review changes" process in GitHub. We can describe the meaning of each option: (for example)

**Comment:** Used to provide non-blocking feedback, such as suggestions, clarifications, or minor edits.
• All comments MUST be considered by the Working Group.
• They MAY be dismissed if deemed non-impactful, already addressed, or out of scope.

**Approve**: Indicates formal agreement with the proposed contribution.
• A minimum of three (3) Working Group members must submit “Approve” reviews for the PR to be eligible for merging.

**Request Changes:** Treated as a strong suggestion to change the contribution.
• If presented as an **OBJECTION**, it MUST be addressed before the contribution can be merged.

mike-finopsorg replied:
@jpradocueva can you make this as a suggested edit please.

shawnalpay replied:
I am concerned about this verbiage here, as we capture the process for review and approval in the Foundation repo.  I want to make sure we don't have duplicative verbiage.  Let's sync on this.

jpradocueva replied:
@shawnalpay, I understand your concern. I rechecked it, and in my opinion, it is still valid to have it here, as this is a guideline on using the three options provided by GitHub, under "Review Changes". I see this as an enhancement of the content in the `foundation` repo, section 4 Pull Request Flow, subsection: Comments.

mike-finopsorg committed:
renamed  spec-development-guidelines.md and small tweaks to development-processes.md

jpradocueva committed:
add clarification on submitting comments

mike-finopsorg committed:
Merge branch 'working_draft' into way_we_work

jpradocueva reviewed (commented):
See comments.

jpradocueva committed:
updates to pull requests review

ljadvey commented on guidelines/development-processes.md:
```diff
@@ -0,0 +1,288 @@
+
+## Git Issues
+
+All work in the FOCUS Project must be tracked via a GitHub Issue. This ensures visibility into what is happening and why changes are being made to the specification. Updates to an Issue's attributes (such as, the due date changing, descriptions, definition of done, assignees) in GitHub should always be accompanied by a comment adding context as to why the change was required. As the group aims to ...
```
```suggestion
All work in the FOCUS Project must be tracked via a GitHub Issue (Work Item, Action Item, etc.). This ensures visibility into what is happening and why changes are being made to the specification. Updates to an Issue's attributes (such as, the due date changing, descriptions, definition of done, assignees) in GitHub should always be accompanied by a comment adding context as to why the change was required. As the group aims to operate more asynchronously these comments are essential.
```

ljadvey commented on guidelines/development-processes.md:
```diff
@@ -0,0 +1,288 @@
+
+### Reviewing
+
+It is important for members to review all PRs especially when they are open and in status of ‘ PR Member Review’ to ensure you have an opportunity to give feedback on all changes to the Specification. Positive feedback is just as important as feedback looking for further changes, without a 👍 or “Looks good to me” comment your support for a PR is not clearly documented.
```
```suggestion
It is important for members to review all PRs especially when they are open and in status of ‘PR Member Review’ to ensure you have an opportunity to give feedback on all changes to the Specification. Positive feedback is just as important as feedback looking for further changes, without a 👍 or “Looks good to me” comment your support for a PR is not clearly documented.
```

ljadvey commented on guidelines/development-processes.md:
```diff
@@ -0,0 +1,288 @@
+
+## Git Notifications
+
+GitHub platform has a [notifications inbox](https://github.com/notifications) which members should regularly check on, in order to keep tabs of updates, mentions, and open assigned work that should be reviewed and responded to. Responding to notifications is important as this enables the group to know that you are aware of and have acknowledged the notification. If you are unable to immediately p ...
```
```suggestion
GitHub platform has a [notifications inbox](https://github.com/notifications) which members should regularly check on, in order to keep tabs of updates, mentions, and open assigned work that should be reviewed and responded to.  Also, members should consider configuring [github notification settings](https://github.com/settings/notifications) and also setting items to be watched so they are notified when changes occur.

Responding to notifications is important as this enables the group to know that you are aware of and have acknowledged the notification. If you are unable to immediately perform the work needed by the notification responding that you have scheduled some time and when to expect an update helps the project keep track of work that is happening.
```

ljadvey commented on guidelines/development-processes.md:
```diff
@@ -0,0 +1,288 @@
+
+#### #task-force-\#
+
+Sub group specific channels for targeted conversations about work items/action items. Remember to update GitHub where relevant.
```
```suggestion
Sub group specific channels for targeted conversations about work items/action items. Remember to update GitHub where relevant.

##### #tf-<name>
Further subgroups are created for larger items like SaaS, sku properties, etc.
```

shawnalpay committed:
Added Project section and Status / AI Status subsection

shawnalpay committed:
Added Content Creation section

mike-finopsorg committed:
Update guidelines/development-processes.md

Co-authored-by: Larry Advey <104788770+ljadvey@users.noreply.github.com>

mike-finopsorg committed:
Update guidelines/development-processes.md

Co-authored-by: Larry Advey <104788770+ljadvey@users.noreply.github.com>

mike-finopsorg committed:
Update guidelines/development-processes.md

Co-authored-by: Larry Advey <104788770+ljadvey@users.noreply.github.com>

shawnalpay commented on guidelines/development-processes.md:
```diff
@@ -0,0 +1,344 @@
+- Under Consideration: being actively evaluated for upcoming spec work
+- Accepted in Scope: will be developed in next release
+- Ready for Dev: not started
+- In Discussion: in progress
```
```suggestion
`In Discovery`: defining the concepts that this issue will entail via Task Force
`In Development`: discussing solution for how to solve for the concept / design via Task Force
```

shawnalpay commented on guidelines/development-tips.md:
Discussed with @mike-finopsorg.  I believe we can now jettison this file, but will leave it to him to take a final review and commit the delete.

jpradocueva commented on guidelines/development-processes.md:
```diff
@@ -0,0 +1,349 @@
+
+All suggestions and feature requests should be introduced to the FOCUS project via GitHub Issues of the type Feature Request. The template for these Issues will ensure the needed information is gathered and available for prioritisation processes. Anyone can submit Feedback Issues within the GitHub repository. The description for feature request items should aim to be clear what the desired outcom ...
+
+Feedback Issues will have the Type: Feature Request and Label: feature request, and be associated with the FOCUS WG project.
```
```suggestion
Feature Request Issues will have the Type: Feature Request and Label: feature request, and be associated with the FOCUS WG project.
```

jpradocueva reviewed (commented):
Suggestion

shawnalpay committed:
Apply suggestions from code review

Co-authored-by: Joaquin <jprado@linuxfoundation.org>

ijurica commented on guidelines/development-processes.md:
```diff
@@ -0,0 +1,350 @@
+
+## Content Creation
+
+This section details best practices for the creation of content that is added to the specification.  This section will be augmented over time.
```
Non-blocking suggestion:

It might be helpful to mention the use of Google Drive during active development - specifically, that it is common practice for each Work Item to have a corresponding Google Drive folder, internally shared among FOCUS members. These folders typically follow a naming convention that includes the Work Item number and a short, human-readable description, making it easier to navigate shared working materials.

mike-finopsorg replied:
@ijurica any chance you have time today/tomorrow to make a suggested edit here?

mike-finopsorg committed:
Merge branch 'working_draft' into way_we_work

mike-finopsorg committed:
removed development-tips.md after discussions

mike-finopsorg commented:
There is a couple of non-blocking ideas that need to be turned into suggestions. I think at this point it would be ok to merge and have these additions come in from their own PRs. Getting this into the repository will be helpful as we move into the next release cycle.

mike-finopsorg committed:
Merge branch 'working_draft' into way_we_work

shawnalpay committed:
Deleted a couple of unnecessary escape slashes

Matt-Cowsert committed:
Update development-processes.md

Added the normative requirements for FR titles and cleaned up a few inconsistencies in the types vs. labels for feature requests and feedback.

Matt-Cowsert commented on guidelines/development-processes.md:
```diff
@@ -0,0 +1,350 @@
+
+Everyone is welcome to add comments to further define the feature request item, add concerns and/or considerations you would like to see taken into account when developing a solution for the feature request. Support for a feature request item can also be shown in the form of reactions on the Issue ticket.
+
+Feature Request Issues will have a title that starts with \`\[FR]\`.
```
Added the normative requirements for Mike to evaluate for inclusion

shawnalpay committed:
Formatting tweak

shawnalpay committed:
Update guidelines/development-processes.md

Co-authored-by: Irena Jurica <irena.jurica@neos.hr>

shawnalpay reviewed (approved):
LGTM!

jpradocueva commented:
These changes were approved during the Members meeting on May 29.
